Youtube scrape transcripts

bszabo opened the pull request:
This PR picks up where [work](https://github.com/openedx/edx-platform/pull/30643) conducted by Crist Ye left off.

Specifically, this feature branch contains modifications to access captions from YouTube videos post-December '21. At or around that time, a non-supported API that had been used to retrieve such captions stopped working. Crist Ye provided a work-around that leverages inclusion of URLs in the video page itself that can then be used to locate the captions.

The files changed here come from two branches in the repo used by Crist Ye's original development. One was his `fix-youtube` branch, holding the scraping logic. The other is the `caption-link-unit-testing branch`, with unit tests for the new logic.

@iamCristYe -- Please have a look here

iamCristYe committed:
fix: importing transcript from YouTube TNL-9460

Commit:
feat: add youtube caption link unit tests

Commit:
feat: TNL-9460 fix regex search from settings


Per 11/10-11/16 exchange in https://github.com/openedx/edx-platform/pull/30643

iamCristYe commented:
Hi @bszabo I'm not sure if you've seen my comments at https://github.com/iamCristYe/edx-platform/pull/1
Others are fine

bszabo commented:
Not all CI checks are passing. I will attend to these a bit later today (Nov 28).

bszabo commented:
@iamCristYe I do not see any comments of yours in https://github.com/iamCristYe/edx-platform/pull/1 (I just double-checked). Any idea why this might be? Is there some change you've recommended?

iamCristYe commented:
Here's my comments regarding line 87 of [xmodule/tests/test_transcripts_utils.py](https://github.com/iamCristYe/edx-platform/pull/1/files/762538da662672a087ec432ce62c57b599ea574f#diff-1153ea5f33fe9fa45fe17177dcbcd92f44d984c865303f98ca42ebad8640b817)

        mock_get.return_value = YoutubeVideoHTMLResponse.with_caption_link(language_code)

        language_specific_caption_link = get_transcript_link_from_youtube(YOUTUBE_VIDEO_ID)
        self.assertEqual(language_specific_caption_link, CAPTION_URL_UTF8_DECODED_TEMPLATE.format(language_code))



I don't know much about this testing framework.. but assertEqual will require the two params to be exactly the same right? But looks like CAPTION_URL_UTF8_DECODED_TEMPLATE have expire time like 1667281544 and also signature that will change every time, so the two params passed in cannot be the same.

mariajgrimaldi reviewed (commented):
Can we not add other types of changes to this PR? This might have happened because of branch conflicts but we need to make sure just the changes related to youtube are added.

bszabo commented:
@mariajgrimaldi Agreed! Thanks for catching this. It was unintentional, and I had not noticed the creep of non-youtube work into the PR. Will try to reverse the unintended merge.

mariajgrimaldi commented:
@bszabo: don't worry! Let me know if you need any help with that.

bszabo commented:
@mariajgrimaldi I believe the unintended merge occurred when I tried to manually resolve conflicts on the GitHub client. I've removed the undesired commit and will look for a cleaner way to resolve the conflicts.

bszabo commented:
@mariajgrimaldi  My efforts to resolve to requirements conflicts are not clearly converging.If you're familiar with how to quickly resolve this type of situation, I'll happily take your advice. Otherwise, I'll just plow on and get there by brute force.

bszabo commented:
I concluded that there have been dependency requirements changes in master that are complicating the task of getting requirements to work properly here, and that it would be best to rebase to master before continuing. I therefore reset to the last meaningful commit, before today's exploratory work with requirements, and will rebase that commit with current master contents.

Commit:
feat: TNL-9460 fix pylint checks



content is occluded
missing docstring

Commit:
feat: TNL-9460 Fix youtube MockResponse


Had occluded content attribute

Commit:
feat: TNL-9460 skip former tests


Old integration tests relying on youtube captions API will no longer work

Commit:
feat: TNL-9460 Mock YOUTUBE env


test_video.py has tests that fail unless YOUTUBE-related environment variables are supplied

bszabo commented:
I discovered yesterday that most of the dependency issues I was encountering had to do with conflicts with recently added upgrades to master. I rebased with master, and the problems cleared up.

There were, however, some legitimate errors dealing with both legacy unit tests and those added with this work. A couple of the most recent commits dealt with my exploration for fixing these.

bszabo commented on xmodule/tests/test_video.py:
```diff
@@ -916,6 +916,12 @@ def test_student_view_data_with_hls_flag(self, mock_get_video_info, mock_get_vid
             'v': 'set_youtube_id_of_11_symbols_here',
         },
     },
+
```
`test_video.py` has legacy tests dealing with youtube transcripts that are complementary to this PR. Because these legacy tests override settings, they need to explicitly offer the updates to the YOUTUBE environment variable added as part of this work to continue working.

Commit:
feat: TNL-9460 unused transcript utility


Remove unused youtube_video_transcript_name(): went with deprecated API

bszabo commented:
@mariajgrimaldi @iamCristYe @connorhaugh The problems I set out to resolve earlier this week have been resolved. I have, however, discovered that unit tests in `cms/djangoapps/contentstore/tests/test_transcript_utils.py` which are worth keeping require back-to-back calls to `requests.get()` in `xmodule/video_module/transcripts_utils.py`, each with different return values. This calls for more elaborate patching than what's currently in place. I'll attend to that next.

Feel free to review what's here now, or you can hold off until the unit test work is done. I'll tag you here when that happens. My goal is to have that done this week.

bszabo commented:
@iamCristYe I have not forgotten about your most-reasonable request to appear as co-author. I do however, need your email to set that up. Please reply here, or email me your address at bszabo@edx.org. Thanks.

Commit:
feat: TNL-9460 debugging multi-patch


For get_transcripts_from_youtube() testing

Commit:
feat: TNL-9460 debugging multi-patch (ii)


Still working with just one test to get patching template right

Commit:
feat: TNL-9460 fix syntax

mariajgrimaldi commented on xmodule/video_module/transcripts_utils.py:
```diff
@@ -166,15 +206,17 @@ def get_transcripts_from_youtube(youtube_id, settings, i18n, youtube_transcript_
+        raise GetTranscriptsFromYouTubeException(msg)
+
+    data = requests.get(transcript_link)
+    log.info('data = {0}'.format(data))
```
why not using f strings?

bszabo replied:
FYI ... This change is throwaway. My devstack environment is acting strangely, and I'm rebuilding it. In the meantime, I'm debugging the failing unit tests by pushing to GitHub. So changes to `transcripts_utils.py` and to the files holding failing unit tests are not keepers.

But your comment is still good, and I'll switch to f strings.

mariajgrimaldi commented on cms/djangoapps/contentstore/tests/test_transcripts_utils.py:
```diff
@@ -287,57 +293,6 @@ def test_success_downloading_chinese_transcripts(self):
-            <track id="1" name="Custom" lang_code="en" />
-            <track id="0" name="Custom1" lang_code="en-GB"/>
-        </transcript_list>
-        """
-        mock_get.return_value = Mock(status_code=200, text=response_success, content=response_success.encode('utf-8'))
-
-        transcript_name = transcripts_utils.youtube_video_transcript_name(youtube_text_api)
-        self.assertEqual(transcript_name, 'Custom')
-
-    @patch('xmodule.video_module.transcripts_utils.requests.get')
-    def test_get_transcript_name_youtube_server_no_transcripts(self, mock_get):
-        """
-        When there are no transcripts of video transcript name will be None
-        """
-        youtube_text_api = copy.deepcopy(settings.YOUTUBE['TEXT_API'])
-        youtube_text_api['params']['v'] = 'dummy_video_id'
-        response_success = "<transcript_list></transcript_list>"
-        mock_get.return_value = Mock(status_code=200, text=response_success, content=response_success.encode('utf-8'))
-
-        transcript_name = transcripts_utils.youtube_video_transcript_name(youtube_text_api)
-        self.assertIsNone(transcript_name)
-
-    @patch('xmodule.video_module.transcripts_utils.requests.get')
-    def test_get_transcript_name_youtube_server_language_not_exist(self, mock_get):
-        """
-        When the language does not exist in transcript_list transcript name will be None
-        """
-        youtube_text_api = copy.deepcopy(settings.YOUTUBE['TEXT_API'])
-        youtube_text_api['params']['v'] = 'dummy_video_id'
-        youtube_text_api['params']['lang'] = 'abc'
-        response_success = """
-        <transcript_list>
-            <track id="1" name="Custom" lang_code="en" />
-            <track id="0" name="Custom1" lang_code="en-GB"/>
-        </transcript_list>
-        """
-        mock_get.return_value = Mock(status_code=200, text=response_success, content=response_success.encode('utf-8'))
-
-        transcript_name = transcripts_utils.youtube_video_transcript_name(youtube_text_api)
-        self.assertIsNone(transcript_name)
```
I suppose these are the legacy tests being translated -if possible- to our current implementation, right?

bszabo replied:
`youtube_video_transcript_name()` has been eliminated, so these tests in their current form do not belong.

Although it's not a one-for-one equivalence, `get_transcript_link_for_youtube()` is covering for what `youtube_video_transcript_name()` used to do. And `get_transcript_link_for_youtube()` is the subject of three new tests in `xmodule/tests/test_transcripts_utils.py`. A case can be made that this file and its counterpart in `cms/djangoapps/contentstore/tests` could be merged there. I'm inclined to leave `xmodule/tests/test_transcripts_utils.py` where it is, as its tests have absolutely nothing to do with `cms` or `contentstore`.

Some additional testing will be required when we bring back support for transcripts in languages other than English. That's out of scope for this PR, however.

mariajgrimaldi commented on xmodule/tests/test_transcripts_utils.py:
```diff
@@ -0,0 +1,103 @@
+
+
+class TranscriptsUtilsTest(TestCase):
+    ''' Tests utility fucntions for transcripts (in video_module)'''
```
we usually use double quotes for docstrings

bszabo replied:
Thanks for calling this out. Makes sense, as paragraph length comments are likely to have contractions with apostrophes in them.

mariajgrimaldi commented on xmodule/tests/test_transcripts_utils.py:
```diff
@@ -0,0 +1,103 @@
+        '''An object fit to be returned from a an HTTP GET operation, exposing
+        a UTF-8 encoded version of the youtube_html input string in its content attribute'''
+        def __init__(self, youtube_html):
+            self.content = bytearray(youtube_html, 'UTF-8')
```
+1! 🥇

mariajgrimaldi commented on xmodule/video_module/transcripts_utils.py:
```diff
@@ -130,27 +131,66 @@ def save_subs_to_store(subs, subs_id, item, language='en'):
+                if caption["languageCode"] == "en":
+                    return caption["baseUrl"]
+        return None
+    except ConnectionError:
```
Should we consider another type of errors? have you encountered any other?

bszabo replied:
The code being replaced here used to have no exception handling at all. I suggest we revisit this question when the unit testing is all done. If we include a test with malformed link info in the HTML and it manifests in some other exception here, we can catch that too.

mariajgrimaldi commented on xmodule/tests/test_transcripts_utils.py:
```diff
@@ -0,0 +1,103 @@
+        mock_get.return_value = YoutubeVideoHTMLResponse.with_no_caption_links()
+
+        english_language_caption_link = get_transcript_link_from_youtube(YOUTUBE_VIDEO_ID)
+        self.assertEqual(english_language_caption_link, None)
```
I have a few questions:
- What happens if the video with YOUTUBE_VIDEO_ID doesn't exist?
- Are we testing what if the API is unavailable? (connection error or status code different than 200?
-  What other cases do you think we're missing?

bszabo replied:
Testing for the defunct `get_transcript_link_for_youtube()` method, which these tests replace, used to cover the following:

- Happy case returns a transcripts URL when one is available for the requested language
- No URL is returned if the video page contains no transcript URLs
- No URL is returned if the video page contains transcript URLs, but not one for the requested language

The tests provided for `get_transcript_link_from_youtube()` cover exactly these three conditions.

Per the previous discussion on possible exceptions we might see, I'll add a test for malformed captions syntax in the link HTML on the video page. If we get such a thing, then the `get_transcript_link_from_youtube()` code may reference the captions dictionary with a missing key;  the test will ensure that nothing bad happens.

iamCristYe commented:
> @iamCristYe I have not forgotten about your most-reasonable request to appear as co-author. I do however, need your email to set that up. Please reply here, or email me your address at [bszabo@edx.org](mailto:bszabo@edx.org). Thanks.

Hi @bszabo It's iamCristYe at Outlook.com

Commit:
feat: TNL-9460 fix contentstore unit tests

Commit:
feat: TNL-9460 fix failing test


contentstore/views test

Commit:
feat: TNL-9460 lint fix

Commit:
feat: TNL-9460 lint fix

bszabo commented:
I’m temporarily suspending this pull request, as the number of commits has become unwieldy, and the last few days have been more of a development phase than a review phase. When the code stabilizes, I will squash commits into a more meaningful package and open a new PR.

The error on the last commit is of a database access nature and quite different from those I was seeing before. I’ll look into it next week, but wouldn’t be surprised if there’s a line of code that was erroneously added or removed in a recent commit.

In the meantime, here’s what I’ve learned.

The former captions solution entailed making two calls to the API: one with no track name supplied, and one with a track name supplied. The first would return a list of all tracks available for a video; the second would return timing information for the selected track.

The current solution likewise entails performing two GET operations, one to get information on all available tracks, and one to get track timing information for one selected track.

On performing a deeper dive on unit tests, I found more unit tests and more utilities than I originally envisioned. With 20-20 hindsight, this is not entirely surprising, given the wide range of content in `transcripts_utils.py`:
  - transcripts can be found in multiple sources (contentsore, blockstore, and YouTube)
  - transcripts have metadata in edx-val
  - transcripts can be encoded as json or as SRT files
Additionally, transcripts can be intended for playback at different speeds, and might be the subject of automatic translation. `transcript_utils.py` sits at the centers of transfers and conversions among these various sources and representations, and is a large file. It would do well to partition it into separate concerns.

  Unit tests are now loosely organized along the following lines:
  - Testing of the first GET operation (the one that obtains a list of all caption tracks) occurs in `xmodule/tests/test_transcripts_utils.py`
  - Testing that involves reading from or saving to video sources (with transcripts going for the ride) resides in `cms/djangopps/contentstore/tests/test_transcript_utils.py` and in `cms/djangoapps/contentstore/views/tests/test_transcripts.py` It is these tests that test the second GET operation (the one that reads timing information for a specific captions track). 

Because these tests make two distinct GET operations, mocking them requires that the return value of the second call be different from the return value of the first call. The majority of late changes pertain to this requirement.

Because the above needs to occur both in `cms/djangoapps/contentstore/test`s and `cms/djangoapps/contentstore/views/tests`, support for double GETs wants to be in a location accessible to both. This is `cms/djangoapps/contentstore/tests/utils.py`
